{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction of N. lanati GEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "include(joinpath(\"..\", \"src\", \"GSM.jl\"))\n",
    "import .GSM\n",
    "using JSON\n",
    "using ExcelReaders\n",
    "using DataValues\n",
    "using BioSequences\n",
    "using DataFrames\n",
    "using StatsBase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = GSM.newmodel(\"iNlan20\") # blank model\n",
    "model[\"version\"] = \"1.0\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load universal model reaction and metabolite data from BIGG\n",
    "db = JSON.parsefile(joinpath(\"suppdata\" ,\"universal_model.json\"))\n",
    "unirxns = GSM.getuniX(db, \"reactions\")\n",
    "unimets = JSON.parsefile(joinpath(\"suppdata\", \"universal_model_metabolites.json\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build draft model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load manually reconstructed model\n",
    "# These include all the primary metabolism reactions\n",
    "# KEGG metabolism maps combined with the master metabolic table of N. lanati were used to identify genes to add to the model\n",
    "ecs, rxnids = GSM.readExcelModel(joinpath(\"suppdata\", \"s3model.xlsx\"), [\"More\",\"Lipids2\", \"Transport2\", \"Transport\", \"Other\",\"Carbohydrate\", \"Amino acids\", \"Exchanges\", \"Nucleotides\", \"Vitamins\", \"Lipids\"])\n",
    "\n",
    "bigg_ec = Dict()\n",
    "for (k, vs) in ecs\n",
    "    if !isna(k)\n",
    "        for v in vs\n",
    "            if haskey(bigg_ec, v)\n",
    "                push!(bigg_ec[v], k)\n",
    "            else\n",
    "                bigg_ec[v] = [k]\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "GSM.fixrxns!(unirxns) # ensure the model is consistent\n",
    "GSM.fixmets!(unimets) # ensure the model is mass and change balanced\n",
    "\n",
    "GSM.fixannotations!(unimets)\n",
    "GSM.fixannotations!(unirxns, bigg_ec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reaction with ID: MI1345PK\n",
      "No reaction with ID: MI145P6K\n"
     ]
    }
   ],
   "source": [
    "# Build model using manually curated data\n",
    "rxnstoadd = String[]\n",
    "metstoadd = String[]\n",
    "for rxnid in rxnids\n",
    "    if haskey(unirxns, rxnid)\n",
    "        push!(rxnstoadd, rxnid)\n",
    "        mets = GSM.getmets(unirxns[rxnid])\n",
    "        if \"h_h\" in mets\n",
    "            println(rxnid)\n",
    "        end\n",
    "        append!(metstoadd, mets)\n",
    "    else\n",
    "        println(\"No reaction with ID: \", rxnid)\n",
    "    end\n",
    "end\n",
    "\n",
    "GSM.addrxns!(model, rxnstoadd, unirxns) # default is reversible\n",
    "GSM.addmets!(model, metstoadd, unimets)\n",
    "\n",
    "include(joinpath(\"..\", \"src\", \"hydrogenosomemodel.jl\")) # add reactions associated with hydrogenosome\n",
    "include(joinpath(\"..\", \"src\", \"add_nonbigg_rxns.jl\")); # add non-BIGG reactions and metabolits to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add biomass objective function\n",
    "atpcoeff = 75.975\n",
    "vitflux = -1e-6 # require vitamin flux\n",
    "bofdict = Dict(\"atp_c\" => -atpcoeff, \"adp_c\" => atpcoeff, # RNA ATP ignore because so small - consumed by maintenance ATP\n",
    "                \"h2o_c\" => -atpcoeff,\n",
    "                \"h_c\" => atpcoeff,\n",
    "                \"pi_c\" => atpcoeff,\n",
    "\n",
    "               \"uacgam_c\" => -0.5157,\n",
    "               \"ttdca_c\" => -0.132,\n",
    "               \"pmtcoa_c\" => -0.132,\n",
    "               \"ocdca_c\" => -0.0297,\n",
    "\n",
    "               \"dgtp_c\" => -0.0020480581936911215,\n",
    "               \"datp_c\" => -0.005646766741414883,\n",
    "               \"dctp_c\" => -0.0017401048397661797,\n",
    "               \"dttp_c\" => -0.004593347112415209,\n",
    "\n",
    "               \"gtp_c\" => -0.00037267485200615695,\n",
    "               \"ctp_c\" => -0.0003737831821377871,\n",
    "               \"utp_c\" => -0.0016833243709196526,\n",
    "\n",
    "               \"ala__L_c\" => -0.144572144215957,\n",
    "               \"ser__L_c\" => -0.326503808396135,\n",
    "               \"gly_c\" => -0.163545035226185,\n",
    "               \"cys__L_c\" => -0.0575278020669201,\n",
    "               \"met__L_c\" => -0.0763784730116249,\n",
    "               \"asn__L_c\" => -0.381782761257429,\n",
    "               \"val__L_c\" => -0.175379485893309,\n",
    "               \"leu__L_c\" => -0.311043758624008,\n",
    "               \"ile__L_c\" => -0.334219837113405,\n",
    "               \"lys__L_c\" => -0.344514314954937,\n",
    "               \"arg__L_c\" => -0.123433592496314,\n",
    "               \"his__L_c\" => -0.0634984493968641,\n",
    "               \"phe__L_c\" => -0.164399787168295,\n",
    "               \"tyr__L_c\" => -0.174586632500512,\n",
    "               \"trp__L_c\" => -0.0304651254938871,\n",
    "               \"pro__L_c\" => -0.133169406360923,\n",
    "               \"thr__L_c\" => -0.208247615578692,\n",
    "               \"asp__L_c\" => -0.219036887556455,\n",
    "               \"glu__L_c\" => -0.28178388067949,\n",
    "               \"gln__L_c\" => -0.124589952019473,\n",
    "\n",
    "               \"thf_c\" => vitflux,\n",
    "               \"ribflv_c\" => vitflux,\n",
    "               \"nad_c\" => vitflux,\n",
    "               \"nadp_c\" => vitflux,\n",
    "               \"pant__R_c\" => vitflux,\n",
    "               \"pydxn_c\" => vitflux,\n",
    "               \"pheme_c\" => vitflux,\n",
    "               \"btn_c\" => vitflux,\n",
    "               \"thm_c\" => vitflux\n",
    "               )\n",
    "\n",
    "GSM.addupbof!(model, bofdict)\n",
    "\n",
    "### add NGAM\n",
    "rxn = Dict()\n",
    "rxn[\"name\"] = string(\"ATPM\")\n",
    "rxn[\"metabolites\"] = Dict(\"atp_c\" => -1,\"h2o_c\" => -1, \"adp_c\" => 1, \"pi_c\" => 1, \"h_c\" => 1)\n",
    "rxn[\"lower_bound\"] = 2.27\n",
    "rxn[\"upper_bound\"] = 2.27\n",
    "rxn[\"id\"] = \"NGAM\"\n",
    "rxn[\"annotation\"] = Dict()\n",
    "rxn[\"gene_reaction_rule\"] = \"\"\n",
    "rxn[\"notes\"] = Dict()\n",
    "GSM.addrxn!(model, rxn);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM.remo2rxns(model) # remove oxygen containing reactions - since N. lanati is anaerobic\n",
    "GSM.remrxns(model, [\"ACS\"]) # remove because no evidence for this in genome\n",
    "GSM.remrxns(model, [\"EX_din_e\", \"FORt\", \"ETOHt\", \"LACDt_c\", \"ACht\", \"FORht\", \"ACtr\"]); # no diffusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of manually checked reaction directions, used metacyc, equilibrator, iMM904 and iJO1366 to constrain reaction directions\n",
    "# if reaction is not listed here it is assumed to be reversible\n",
    "open(joinpath(\"suppdata\",\"iNlan20_rxn_directions.tab\")) do io \n",
    "    firstline=true\n",
    "    for ln in eachline(io)\n",
    "        firstline && (firstline=false; continue)\n",
    "        prts = split(ln, \"\\t\")\n",
    "        id = prts[1]\n",
    "        lb = parse(Float64, prts[2])\n",
    "        ub = parse(Float64, prts[3])\n",
    "        GSM.changebound!(model, id, lb, ub)\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign gene reaction rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "bidir_ec_pids = Dict()\n",
    "jgi_ec_pids = Dict()\n",
    "open(joinpath(\"..\", \"MetabolicTables\", \"Neosp3.tab\")) do io\n",
    "    firstline = true\n",
    "    for ln in eachline(io)\n",
    "        firstline && (firstline=false; continue)\n",
    "        prts = split(ln, \"\\t\")\n",
    "        bidir_ecs = split(prts[5], \"; \")\n",
    "        jgi_ecs = split(prts[4], \"; \")\n",
    "        pid = prts[1]\n",
    "        \n",
    "        for ec in unique(bidir_ecs)\n",
    "           bidir_ec_pids[ec] = push!(get(bidir_ec_pids, ec, String[]), pid) \n",
    "        end\n",
    "        for ec in unique(jgi_ecs)\n",
    "           jgi_ec_pids[ec] = push!(get(jgi_ec_pids, ec, String[]), pid)\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "406\n",
      "406\n"
     ]
    }
   ],
   "source": [
    "# Add bidirectional annotations\n",
    "allpids = String[]\n",
    "for k=1:length(model[\"reactions\"])\n",
    "    modelecs = [string(x) for x in get(model[\"reactions\"][k][\"annotation\"], \"ec-code\", [])]\n",
    "    gprs = String[]\n",
    "    found = false\n",
    "    for modelec in modelecs\n",
    "        pids = [string(x) for x  in get(bidir_ec_pids, modelec, [])]\n",
    "        if !isempty(pids)\n",
    "            found = true\n",
    "            for pid in pids\n",
    "                push!(gprs, pid)\n",
    "                push!(allpids, pid)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    if found\n",
    "        model[\"reactions\"][k][\"gene_reaction_rule\"] = join(unique(gprs), \" or \")\n",
    "        model[\"reactions\"][k][\"notes\"][\"Annotation_source\"] = \"Bidirectional blasting\"\n",
    "    end\n",
    "end\n",
    "\n",
    "k = 0\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        gene = Dict()\n",
    "        k += 1\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"Bidirectional blasting\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71\n",
      "477\n"
     ]
    }
   ],
   "source": [
    "# Add JGI annotations\n",
    "allpids = String[]\n",
    "for k=1:length(model[\"reactions\"])\n",
    "    modelecs = [string(x) for x in get(model[\"reactions\"][k][\"annotation\"], \"ec-code\", [])]\n",
    "    gprs = String[]\n",
    "    found = false\n",
    "    for modelec in modelecs\n",
    "        pids = [string(x) for x  in get(jgi_ec_pids, modelec, [])]\n",
    "        if !isempty(pids)\n",
    "            found = true\n",
    "            for pid in pids\n",
    "                push!(gprs, pid)\n",
    "                push!(allpids, pid)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    if found\n",
    "        model[\"reactions\"][k][\"gene_reaction_rule\"] = join(unique(gprs), \" or \")\n",
    "        model[\"reactions\"][k][\"notes\"][\"Annotation_source\"] = \"JGI\"\n",
    "    end\n",
    "end\n",
    "k = 0\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        k += 1\n",
    "        gene = Dict()\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"JGI\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "451\n",
      "927\n"
     ]
    }
   ],
   "source": [
    "# Read cellulases and hemicellulases from JGI cazyme annotation data\n",
    "cazy_pids = Dict() # r_cellulase, r_hemicellulase\n",
    "cellulases, hemicellulases = GSM.getcazygenes(joinpath(\"..\", \"OmicsData\", \"Cazymes\", \"cazyreport.txt\"))\n",
    "cellulase_rxn = GSM.getrxn(model, \"r_cellulase\")\n",
    "cellulase_rxn[\"gene_reaction_rule\"] = join(unique(cellulases), \" or \")\n",
    "cellulase_rxn[\"notes\"][\"Annotation_source\"] = \"JGI\"\n",
    "hemicellulase_rxn = GSM.getrxn(model, \"r_hemicellulase\")\n",
    "hemicellulase_rxn[\"gene_reaction_rule\"] = join(unique(hemicellulases), \" or \")\n",
    "hemicellulase_rxn[\"notes\"][\"Annotation_source\"] = \"JGI\" \n",
    "\n",
    "allpids = [cellulases; hemicellulases]\n",
    "k = 1\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        k += 1\n",
    "        gene = Dict()\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"JGI\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      "949\n"
     ]
    }
   ],
   "source": [
    "# Add genes based on homology to iMM904 genes (bidirectional)\n",
    "# add note indicating these genes are based on homology to S. cerevisiae\n",
    "fungus_to_other_loc = joinpath(\"..\", \"OmicsData\", \"Yeast\", \"neosp3.yeast.out\")\n",
    "other_to_fungus_loc = joinpath(\"..\", \"OmicsData\", \"Yeast\", \"yeast.neosp3.out\")\n",
    "bidir = GSM.reversebidir(GSM.matchbidir(fungus_to_other_loc, other_to_fungus_loc, 1e-20, 3))\n",
    "uniprot = GSM.readuniprot(joinpath(\"..\", \"OmicsData\", \"Yeast\", \"yeast_genes.tab\"), [1,5])\n",
    "genenames_uniprot = GSM.genes_uniprot(uniprot)\n",
    "yeastmodel = GSM.loadmodel(joinpath(\"..\", \"OmicsData\", \"Yeast\", \"iMM904.json\"))\n",
    "\n",
    "allpids = String[]\n",
    "yrxn_genes = Dict()\n",
    "for rxn in yeastmodel[\"reactions\"]\n",
    "    id = rxn[\"id\"]\n",
    "    mrxn = GSM.getrxn(model, id)\n",
    "    isempty(mrxn) && continue\n",
    "    (mrxn[\"gene_reaction_rule\"] != \"\") && continue\n",
    "    # only add to rxns if not already annotated\n",
    "    genesor = split(rxn[\"gene_reaction_rule\"], \" or \")\n",
    "    genesand = split(rxn[\"gene_reaction_rule\"], \" and \")\n",
    "    genes = unique([genesor; genesand])\n",
    "    yrxn_genes[id] = genes\n",
    "    grr = String[]\n",
    "    found = false\n",
    "    for gene in genes\n",
    "        uid = get(genenames_uniprot, gene, \"X\")\n",
    "        pid = get(bidir, uid, \"\")\n",
    "        if pid != \"\"\n",
    "            push!(grr, pid)\n",
    "            found = true\n",
    "        end\n",
    "    end\n",
    "    if found\n",
    "        mrxn[\"gene_reaction_rule\"] = join(unique(grr), \" or \")\n",
    "        mrxn[\"notes\"][\"Annotation_source\"] = \"S. cerevisiae GEM\"\n",
    "        append!(allpids, grr)\n",
    "    end\n",
    "end\n",
    "\n",
    "k = 0\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        k += 1\n",
    "        gene = Dict()\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"S. cerevisiae (iMM904)\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "952\n"
     ]
    }
   ],
   "source": [
    "# Add genes based on homology to iML1515 genes (bidirectional)\n",
    "# add note indicating these genes are based on homology to E. coli\n",
    "fungus_to_other_loc = joinpath(\"..\", \"OmicsData\", \"Ecoli\", \"neosp3.ecoli.out\")\n",
    "other_to_fungus_loc = joinpath(\"..\", \"OmicsData\", \"Ecoli\", \"ecoli.neosp3.out\")\n",
    "bidir = GSM.reversebidir(GSM.matchbidir(fungus_to_other_loc, other_to_fungus_loc, 1e-20, 3))\n",
    "uniprot = GSM.readuniprot(joinpath(\"..\", \"OmicsData\", \"Ecoli\", \"ecoli.tab\"), [1,5])\n",
    "genenames_uniprot = GSM.genes_uniprot(uniprot)\n",
    "ecolimodel = GSM.loadmodel(joinpath(\"..\", \"OmicsData\", \"Ecoli\", \"iML1515.json\"))\n",
    "\n",
    "allpids = String[]\n",
    "for rxn in ecolimodel[\"reactions\"]\n",
    "    id = rxn[\"id\"]\n",
    "    mrxn = GSM.getrxn(model, id)\n",
    "    isempty(mrxn) && continue\n",
    "    (mrxn[\"gene_reaction_rule\"] != \"\") && continue\n",
    "    # only add to rxns if not already annotated\n",
    "    genesor = split(rxn[\"gene_reaction_rule\"], \" or \")\n",
    "    genesand = split(rxn[\"gene_reaction_rule\"], \" and \")\n",
    "    genes = unique([genesor; genesand])\n",
    "    grr = String[]\n",
    "    found = false\n",
    "    for gene in genes\n",
    "        uid = get(genenames_uniprot, gene, \"X\")\n",
    "        pid = get(bidir, uid, \"\")\n",
    "        if pid != \"\"\n",
    "            push!(grr, pid)\n",
    "            found = true\n",
    "        end\n",
    "    end\n",
    "    if found\n",
    "        mrxn[\"gene_reaction_rule\"] = join(unique(grr), \" or \")\n",
    "        mrxn[\"notes\"][\"Annotation_source\"] = \"E. coli GEM\"\n",
    "        append!(allpids, grr)\n",
    "    end\n",
    "end\n",
    "\n",
    "k = 0\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        k += 1\n",
    "        gene = Dict()\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"E. coli (iML1515)\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66\n",
      "1018\n"
     ]
    }
   ],
   "source": [
    "# read file of manually annotated genes combined from many sources\n",
    "allpids = String[]\n",
    "open(joinpath(\"suppdata\", \"grr.tab\")) do io\n",
    "   for ln in eachline(io)\n",
    "        prts = split(ln, \"\\t\")\n",
    "        rxnid = prts[1]\n",
    "        grr = prts[2]\n",
    "        if grr == \"\"\n",
    "            continue\n",
    "        end\n",
    "        homo = prts[3]\n",
    "        rxn = GSM.getrxn(model, rxnid)\n",
    "        rxn[\"gene_reaction_rule\"] = grr\n",
    "        rxn[\"notes\"][\"Annotation_source\"] = \"Manual\"\n",
    "        if homo != \"None\"\n",
    "            rxn[\"notes\"][\"Homology_genes\"] = homo\n",
    "        end\n",
    "        append!(allpids, split(replace(replace(prts[2], \" or \"=>\" \"), \" and \" => \" \"), \" \"))\n",
    "    end\n",
    "end\n",
    "k = 0\n",
    "for pid in unique(allpids)\n",
    "    if isempty(GSM.getgene(model, pid))\n",
    "        k += 1\n",
    "        gene = Dict()\n",
    "        gene[\"id\"] = pid\n",
    "        gene[\"name\"] = \"Protein ID: $(pid)\"\n",
    "        gene[\"notes\"] = Dict(\"Annotation source\" => \"Manual curation\")\n",
    "        gene[\"annotation\"] = Dict(\"sbo\" => [\"SBO:0000243\"], \"jgi-protein-id\" => pid)\n",
    "        push!(model[\"genes\"], gene)\n",
    "    end\n",
    "end\n",
    "println(k)\n",
    "println(length(model[\"genes\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add additional information\n",
    "Subsystem, confidence score etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "More\n",
      "Lipids2\n",
      "Transport2\n",
      "Transport\n",
      "Other\n",
      "Carbohydrate\n",
      "Amino acids\n",
      "Exchanges\n",
      "Nucleotides\n",
      "Vitamins\n",
      "Lipids\n",
      "sew rxns\n"
     ]
    }
   ],
   "source": [
    "# Add subsystem information\n",
    "sheets = [\"More\",\"Lipids2\", \"Transport2\", \"Transport\", \"Other\",\"Carbohydrate\", \"Amino acids\", \"Exchanges\", \"Nucleotides\", \"Vitamins\", \"Lipids\", \"sew rxns\"]\n",
    "biggid_subs = Dict()\n",
    "for sheet in sheets\n",
    "    println(sheet)\n",
    "    xl = readxlsheet(joinpath(\"suppdata\", \"s3model.xlsx\"), sheet)\n",
    "    for i=2:size(xl, 1)\n",
    "        if !isna(xl[i, 2])\n",
    "            biggid_subs[xl[i, 2]] = xl[i, 4]\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "hyds = [\"H2ht\", \"CO2ht\", \"PIht\", \"H2Oht\", \"ADPATPht\",\"SUCCht1\", \"SUCCht2\", \"MALht\", \"FORht\", \"FORht2\",\"ACht\",\"ACht2\",\"PFLh\",\"PFOh\", \"HYDhfe\",\"CMPL1h\",\"CMPL2h\",\"FUMh\", \"HYDhbi\",\"HYDhfe\",\"ATPShydr\",\"SUCCOAh\",\"SUCOAACTh\",\"MEhy\",\"MEhx\", \"MDHh\", \"AMTh\", \"MAKht\", \"AGht\", \"MASh\"]\n",
    "for hyd in hyds\n",
    "    biggid_subs[hyd] = \"Hydrogenosome\"\n",
    "end\n",
    "for rxn in model[\"reactions\"]\n",
    "    id = rxn[\"id\"]\n",
    "    if haskey(rxn, \"notes\")\n",
    "        rxn[\"notes\"][\"Subsystem\"] = get(biggid_subs, id, \"\")\n",
    "    else\n",
    "        rxn[\"notes\"] = Dict()\n",
    "        rxn[\"notes\"][\"Subsystem\"] = get(biggid_subs, id, \"\")\n",
    "    end\n",
    "end\n",
    "\n",
    "# Add confidence information\n",
    "# Levels: 0 = gap, 1 = inferred from other gut fungi, 2 = Gene, 3 = Gene + transcript\n",
    "anasp1 = GSM.readecs(joinpath(\"suppdata\", \"Anasp1.tab\"))\n",
    "neosp1 = GSM.readecs(joinpath(\"suppdata\", \"Neosp1.tab\"))\n",
    "pirfi3 = GSM.readecs(joinpath(\"suppdata\", \"Pirfi3.tab\"))\n",
    "pire2 = GSM.readecs(joinpath(\"suppdata\", \"PirE2.tab\"))\n",
    "orpsp1 = GSM.readecs(joinpath(\"suppdata\", \"Orpsp1.tab\"))\n",
    "\n",
    "transcriptomicsbase = joinpath(\"..\", \"OmicsData\", \"Transcriptomics\", \"Results\")\n",
    "pid_tid = Dict()\n",
    "open(joinpath(transcriptomicsbase, \"Neosp3.out\")) do f\n",
    "    for ln in eachline(f)\n",
    "        prts = split(ln, \"\\t\")\n",
    "        pid = prts[2]\n",
    "        tid = prts[1]\n",
    "        tparts = join(split(tid, \"_\")[1:end-1], \"_\")\n",
    "        eval = parse(Float64, prts[3])\n",
    "        if !haskey(pid_tid, pid) && eval < 1e-20 # first hit is the best alignment\n",
    "            pid_tid[pid] = tparts\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "score_counts = Dict()\n",
    "helper_anno_fungi = Dict()\n",
    "rxn_fungi = Dict()\n",
    "for rxn in model[\"reactions\"]\n",
    "    # println(rxn[\"id\"])\n",
    "    confidence_score = \"\"\n",
    "    if rxn[\"gene_reaction_rule\"] == \"\" && haskey(rxn, \"annotation\")\n",
    "        fungalhits = String[]\n",
    "        if haskey(rxn[\"annotation\"], \"ec-code\")\n",
    "            for ec in rxn[\"annotation\"][\"ec-code\"]\n",
    "                if ec in anasp1\n",
    "                    push!(fungalhits, \"A. robustus\")\n",
    "                end\n",
    "                if ec in neosp1\n",
    "                    push!(fungalhits, \"N. californiae\")\n",
    "                end\n",
    "                if ec in pirfi3\n",
    "                    push!(fungalhits, \"P. finnis\")\n",
    "                end\n",
    "                if ec in pire2\n",
    "                    push!(fungalhits, \"P. sp. E2\")\n",
    "                end\n",
    "                if ec in orpsp1\n",
    "                    push!(fungalhits, \"P. ruminantium\")\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "\n",
    "        if !isempty(fungalhits)\n",
    "            for fungus in fungalhits\n",
    "                helper_anno_fungi[fungus] = get(helper_anno_fungi, fungus, 0) + 1\n",
    "            end\n",
    "            confidence_score = string(\"1 (\", join(fungalhits,\", \") ,\")\")\n",
    "            rxn_fungi[rxn[\"id\"]] = fungalhits\n",
    "        end\n",
    "\n",
    "        if confidence_score == \"\"\n",
    "            confidence_score = \"0\"\n",
    "        end\n",
    "    else\n",
    "        if occursin(\" or \", rxn[\"gene_reaction_rule\"])\n",
    "            genes = split(rxn[\"gene_reaction_rule\"], \" or \")\n",
    "        elseif occursin(\" and \", rxn[\"gene_reaction_rule\"])\n",
    "            genes = split(rxn[\"gene_reaction_rule\"], \" and \")\n",
    "        else\n",
    "            genes = [rxn[\"gene_reaction_rule\"]]\n",
    "        end\n",
    "        for gene in genes\n",
    "            if haskey(pid_tid, gene)\n",
    "                confidence_score = \"3\"\n",
    "                break\n",
    "            end\n",
    "        end\n",
    "        if confidence_score != \"3\"\n",
    "            confidence_score = \"2\"\n",
    "        end\n",
    "    end\n",
    "\n",
    "    if confidence_score == \"0\"\n",
    "        # excludes exchange reactions\n",
    "        if !occursin(\"EX_\", rxn[\"id\"])\n",
    "            score_counts[\"0\"] = get(score_counts, \"0\", 0) + 1\n",
    "        end\n",
    "    elseif confidence_score == \"2\"\n",
    "        score_counts[\"2\"] = get(score_counts, \"2\", 0) + 1\n",
    "    elseif confidence_score == \"3\"\n",
    "        score_counts[\"3\"] = get(score_counts, \"3\", 0) + 1\n",
    "    else\n",
    "        score_counts[\"1\"] = get(score_counts, \"1\", 0) + 1\n",
    "    end\n",
    "    rxn[\"notes\"][\"Confidence_score\"] = confidence_score\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM.fixmisc(model);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM.savemodel(model, \"iNlan20.json\") # this will not work yet, you need to open it with cobrapy and save it from there"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now convert this model to an SBML model using cobrapy\n",
    "Use python script \"Write model.ipynb\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
